src/cryptsetup/cryptsetup-tpm2.c

```diff
@@ -1,14 +1,63 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
 #include "alloc-util.h"
+#include "ask-password-api.h"
 #include "cryptsetup-tpm2.h"
+#include "env-util.h"
 #include "fileio.h"
 #include "hexdecoct.h"
 #include "json.h"
 #include "parse-util.h"
 #include "random-util.h"
 #include "tpm2-util.h"
 
+static int get_pin(usec_t until, AskPasswordFlags ask_password_flags, bool headless, char **ret_pin_str) {
+        _cleanup_free_ char *pin_str = NULL;
+        _cleanup_strv_free_erase_ char **pin = NULL;
+        int r;
+
+        assert(ret_pin_str);
+
+        r = getenv_steal_erase("PIN", &pin_str);
+        if (r < 0)
+                return log_error_errno(r, "Failed to acquire PIN from environment: %m");
+        if (!r) {
+                if (headless)
+                        return log_error_errno(
+                                        SYNTHETIC_ERRNO(ENOPKG),
+                                        "PIN querying disabled via 'headless' option. "
+                                        "Use the '$PIN' environment variable.");
+
+                pin = strv_free_erase(pin);
+                r = ask_password_auto(
+                        "Please enter TPM2 PIN:",
+                        "drive-harddisk",
+                        NULL,
+                        "tpm2-pin",
+                        "cryptsetup.tpm2-pin",
+                        until,
+                        ask_password_flags,
+                        &pin);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to ask for user pin: %m");
+                assert(strv_length(pin) == 1);
+
+                if (!tpm2_valid_pin(*pin)) {
```

Comment:
doesn't matter, but given this in theory is an array of strings, I'd access the first element as `pin[0]` rather than `*pin`, just as a hint to the reader that that's what we do here: not just deref a pointer, but get the first element from an array the pointer points too, even if technically both are the same thing.